### ISSUE TITLE
Installing tensorflow go failed

### ISSUE THREAD

**Issue by freemanliu** [tensorflow/tensorflow#35133]
<em>Please make sure that this is a build/installation issue. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:build_template</em>

**System information**
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): mac os 10.14.6
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:
- TensorFlow installed from (source or binary): n/a
- TensorFlow version: n/a
- Python version: n/a
- Installed using virtualenv? pip? conda?: n/a
- Bazel version (if compiling from source): n/a
- GCC/Compiler version (if compiling from source): n/a
- CUDA/cuDNN version: n/a
- GPU model and memory: n/a

**Describe the problem**
Followed https://www.tensorflow.org/install/lang_go
but got this:
go get github.com/tensorflow/tensorflow/tensorflow/go
package github.com/tensorflow/tensorflow/tensorflow/go/genop/internal/proto/github.com/tensorflow/tensorflow/tensorflow/go/core: cannot find package "github.com/tensorflow/tensorflow/tensorflow/go/genop/internal/proto/github.com/tensorflow/tensorflow/tensorflow/go/core" in any of:

go get -d github.com/tensorflow/tensorflow/tensorflow/go
yielded the same result.

**Any other info / logs**
Include any logs or source code that would be helpful to diagnose the problem. If including tracebacks, please include the full traceback. Large logs and files should be attached.

**Comment by ymodak** [comment]
I think there is a typo in your ```get``` command there isn't package word in the end.
Can you try again with,
```go get github.com/tensorflow/tensorflow/tensorflow/go```

**Comment by freemanliu** [comment]
hi, ymodak,

Thanks for the quick response!

The "package" is the starting of the output of the command.

C library is installed. Here is how it works:
$gcc hello_tf.c -ltensorflow -o hello_tf
$./hello_tf 
Hello from TensorFlow C library version 1.15.0

I also tried on linux and the result is same.

Neither does building from source work.

I also checked github for 
"github.com/tensorflow/tensorflow/tensorflow/go/genop/internal/proto/github.com/tensorflow/tensorflow/tensorflow/go/core". But it stopped at
https://github.com/tensorflow/tensorflow/tree/master/tensorflow/go/genop/internal.

Is the left part generated automatically? If so, how?

Cheers,
Freeman

**Comment by freemanliu** [comment]
Suspecting my system is in a bad state, I started a tensorflow container and tried to install in it. And it failed in the exact same way. I started to doubt that the whole thing is broken. Can someone spend 10 minutes to try to install it to confirm that it works? (on the other hand, it is really hard to believe that this is broken in such a way) search on google and most of result are from 2 years ago.

**Comment by freemanliu** [comment]
Otherwise I have to switch to the tensorflow java.

**Comment by freemanliu** [comment]
saw https://github.com/tensorflow/tensorflow/issues/23257. Not exactly the same issue. But the common part is that installing tensorflow for go fails. There should be an integration test for this.

**Comment by nphase** [comment]
I have this same issue:

```➜  tf gcc hello_tf.c -ltensorflow -o hello_tf
➜  tf ./hello_tf
Hello from TensorFlow C library version 1.15.0
➜  tf cd
➜  ~ go get github.com/tensorflow/tensorflow/tensorflow/go
package github.com/tensorflow/tensorflow/tensorflow/go/genop/internal/proto/github.com/tensorflow/tensorflow/tensorflow/go/core: cannot find package "github.com/tensorflow/tensorflow/tensorflow/go/genop/internal/proto/github.com/tensorflow/tensorflow/tensorflow/go/core" in any of:
	/usr/local/Cellar/go@1.12/1.12.13/libexec/src/github.com/tensorflow/tensorflow/tensorflow/go/genop/internal/proto/github.com/tensorflow/tensorflow/tensorflow/go/core (from $GOROOT)
	/Users/wschobeiri/go/src/github.com/tensorflow/tensorflow/tensorflow/go/genop/internal/proto/github.com/tensorflow/tensorflow/tensorflow/go/core (from $GOPATH)

**Comment by nphase** [comment]
Checking out `36ef4264270e3815ea4b485a52e35e26e43e2112` per the other thread is a viable workaround.

**Comment by freemanliu** [comment]
Another workaround is Java. It is old and not as fun to work with as go.
But kotlin comes to help. Actually I found kotlin has a few features making
it even more fun to work with than go, e.g., null handling, exception, and
of course generic. Besides, I can use Gradle to manage packages and
dependencies. Everything works smoothly so far.

**Comment by nphase** [comment]
> Another workaround is Java. It is old and not as fun to work with as go. But kotlin comes to help. Actually I found kotlin has a few features making it even more fun to work with than go, e.g., null handling, exception, and of course generic. Besides, I can use Gradle to manage packages and dependencies. Everything works smoothly so far.
> […](#)

Unfortunately rewriting an entire application into another language is not the most convenient option..

**Comment by freemanliu** [comment]
Yes, I just started. Tf go should have some basic integration test.

**Comment by mmussett** [comment]
Same issue with me....

go get github.com/tensorflow/tensorflow/tensorflow/go
package github.com/tensorflow/tensorflow/tensorflow/go/genop/internal/proto/github.com/tensorflow/tensorflow/tensorflow/go/core: cannot find package "github.com/tensorflow/tensorflow/tensorflow/go/genop/internal/proto/github.com/tensorflow/tensorflow/tensorflow/go/core" in any of:
	/usr/local/Cellar/go/1.13.6/libexec/src/github.com/tensorflow/tensorflow/tensorflow/go/genop/internal/proto/github.com/tensorflow/tensorflow/tensorflow/go/core (from $GOROOT)
	/Users/mmussett/src/go/src/github.com/tensorflow/tensorflow/tensorflow/go/genop/internal/proto/github.com/tensorflow/tensorflow/tensorflow/go/core (from $GOPATH)

Tried go clean

**Comment by mr-martin-metal** [comment]
Same issue. C-API in place, tests OK. go 1.11:

go get github.com/tensorflow/tensorflow/tensorflow/go
package github.com/tensorflow/tensorflow/tensorflow/go/genop/internal/proto/github.com/tensorflow/tensorflow/tensorflow/go/core: cannot find package "github.com/tensorflow/tensorflow/tensorflow/go/genop/internal/proto/github.com/tensorflow/tensorflow/tensorflow/go/core" in any of:
	/opt/go/1.11.2/src/github.com/tensorflow/tensorflow/tensorflow/go/genop/internal/proto/github.com/tensorflow/tensorflow/tensorflow/go/core (from $GOROOT)
	/home/martinme/Develop/go/src/github.com/tensorflow/tensorflow/tensorflow/go/genop/internal/proto/github.com/tensorflow/tensorflow/tensorflow/go/core (from $GOPATH)

**Comment by NickvdMeij** [comment]
same here using docker.

`Sending build context to Docker daemon  11.26kB
Step 1/6 : FROM golang:1.13
 ---> 272e3f68338f
Step 2/6 : WORKDIR /go/src/app
 ---> Using cache
 ---> ab323f30415a
Step 3/6 : COPY ./go .
 ---> Using cache
 ---> ceb815a86117
Step 4/6 : RUN go get -d -v github.com/tensorflow/tensorflow/tensorflow/go
 ---> Running in 366c170d1ab3
github.com/tensorflow/tensorflow (download)
github.com/golang/protobuf (download)
package github.com/tensorflow/tensorflow/tensorflow/go/genop/internal/proto/github.com/tensorflow/tensorflow/tensorflow/go/core: cannot find package "github.com/tensorflow/tensorflow/tensorflow/go/genop/internal/proto/github.com/tensorflow/tensorflow/tensorflow/go/core" in any of:
        /usr/local/go/src/github.com/tensorflow/tensorflow/tensorflow/go/genop/internal/proto/github.com/tensorflow/tensorflow/tensorflow/go/core (from $GOROOT)
        /go/src/github.com/tensorflow/tensorflow/tensorflow/go/genop/internal/proto/github.com/tensorflow/tensorflow/tensorflow/go/core (from $GOPATH)
The command '/bin/sh -c go get -d -v github.com/tensorflow/tensorflow/tensorflow/go' returned a non-zero code: 1`

dockerfile:
`FROM golang:1.13

WORKDIR /go/src/app
COPY ./go .

RUN go get -d -v github.com/tensorflow/tensorflow/tensorflow/go
RUN go install -v github.com/tensorflow/tensorflow/tensorflow/go

CMD ["app"]`

**Comment by jhseu** [comment]
Reopening since there's still an issue. Will resolve tomorrow.

**Comment by jhseu** [comment]
Fixed as of https://github.com/tensorflow/tensorflow/commit/4221d1aa4d20ada495771528bb13ca786d0bdbe0

**Comment by jhseu** [comment]
(As a sidenote, please follow the instructions on the Go README which differs slightly from the documentation on tensorflow.org. I'll work on updating the official docs):
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/go/README.md

**Comment by glarchev** [comment]
Issue is not fixed:

`go get github.com/tensorflow/tensorflow/tensorflow/go
package github.com/tensorflow/tensorflow/tensorflow/go/core/core_protos_go_proto: cannot find package "github.com/tensorflow/tensorflow/tensorflow/go/core/core_protos_go_proto" in any of:
	/usr/local/go/src/github.com/tensorflow/tensorflow/tensorflow/go/core/core_protos_go_proto (from $GOROOT)
       /Users/admin/go/src/github.com/tensorflow/tensorflow/tensorflow/go/core/core_protos_go_proto (from $GOPATH)`

**Comment by jhseu** [comment]
@glarchev Please follow the instructions in the README:
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/go/README.md

You need to use `go get -d` to download-only and then use go generate to create the protobuf files.

**Comment by yutkin** [comment]
The same issue:

```
RUN go get -d -u -v github.com/tensorflow/tensorflow/tensorflow/go
 ---> Running in 52496c3a3564
github.com/tensorflow/tensorflow (download)
created GOPATH=/root/go; see 'go help gopath'
github.com/golang/protobuf (download)
package github.com/tensorflow/tensorflow/tensorflow/go/core/core_protos_go_proto: cannot find package "github.com/tensorflow/tensorflow/tensorflow/go/core/core_protos_go_proto" in any of:
	/go/src/github.com/tensorflow/tensorflow/tensorflow/go/core/core_protos_go_proto (from $GOROOT)
	/root/go/src/github.com/tensorflow/tensorflow/tensorflow/go/core/core_protos_go_proto (from $GOPATH)
The command '/bin/sh -c go get -d -u -v github.com/tensorflow/tensorflow/tensorflow/go' returned a non-zero code: 1
```

**Comment by jhseu** [comment]
You can actually ignore that error for now and run:
```
$ go generate github.com/tensorflow/tensorflow/tensorflow/go/op
$ go test github.com/tensorflow/tensorflow/tensorflow/go
ok  	github.com/tensorflow/tensorflow/tensorflow/go	0.182s
```

**Comment by yutkin** [comment]
I cannot ignore it, because `go get -d github.com/tensorflow/tensorflow/tensorflow/go` fails with exit code 1, that interrupts build of Docker image.

**Comment by glarchev** [comment]
The installation steps in README also fail on the `./configure` step if you don't have bazel version 2.0.0 (I have 2.1.0).

**Comment by Mathisco-01** [comment]
I know #36523 is still open and waiting for review, but I just wanted to bump.

**Comment by Pisush** [comment]
I'm with 
```
$ go version 
go1.14.2 darwin/amd64
```
And
```
$ bazel version
Build label: 2.0.0
Build target: bazel-out/darwin-opt/bin/src/main/java/com/google/devtools/build/lib/bazel/BazelServer_deploy.jar
Build time: Thu Dec 19 12:33:30 2019 (1576758810)
Build timestamp: 1576758810
Build timestamp as int: 1576758810
```
Installed this version by removing current bazel using brew and running
`$ brew install https://raw.githubusercontent.com/bazelbuild/homebrew-tap/1bd39af61f9731fd03ba4dde408a0cb4198bc76d/Formula/bazel.rb`

Then I tried running my code again:
```
$ go run main.go
go: finding module for package github.com/tensorflow/tensorflow/tensorflow/go/core/core_protos_go_proto
(in the first attempt:) go: downloading github.com/tensorflow/tensorflow v1.15.2
../../../go/pkg/mod/github.com/tensorflow/tensorflow@v2.2.0+incompatible/tensorflow/go/saved_model.go:24:2: module github.com/tensorflow/tensorflow@latest found (v2.2.0+incompatible), but does not contain package github.com/tensorflow/tensorflow/tensorflow/go/core/core_protos_go_proto
```

Then switched to following go [Go installation instructions](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/go/README.md):
```
$ go get -d github.com/tensorflow/tensorflow/tensorflow/go
go: found github.com/tensorflow/tensorflow/tensorflow/go in github.com/tensorflow/tensorflow v2.2.0+incompatible
go: finding module for package github.com/tensorflow/tensorflow/tensorflow/go/core/core_protos_go_proto
```

```
$cd ${GOPATH}/src/github.com/tensorflow/tensorflow
$./configure
Traceback (most recent call last):
  File "./configure.py", line 1549, in <module>
    main()
  File "./configure.py", line 1365, in main
    _TF_MAX_BAZEL_VERSION)
  File "./configure.py", line 478, in check_bazel_version
    ['bazel', '--batch', '--bazelrc=/dev/null', 'version'])
  File "./configure.py", line 156, in run_shell
    output = subprocess.check_output(cmd)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/subprocess.py", line 573, in check_output
    raise CalledProcessError(retcode, cmd, output=output)
subprocess.CalledProcessError: Command '['bazel', '--batch', '--bazelrc=/dev/null', 'version']' returned non-zero exit status 1
```

People in the thread seem to have succeeded with the bazel v2.0.0 but this doesn't seem to work for me. So bumping as well.

**Comment by atahmasb** [comment]
you need to checkout to r1.11 branch, as explained [here](https://github.com/tensorflow/tensorflow/issues/23257#issuecomment-433751410), or you can install the master plus the latest version of the C library (1.15) and in your `go.mod` under `require` change the package version to `github.com/tensorflow/tensorflow v1.11.0` so that you wouldn't have to checkout to that branch every time you import the package.

**Comment by galeone** [comment]
Since this issue isn't getting fixed, I decided to maintain a fork: https://github.com/galeone/tensorflow

```
go get github.com/galeone/tensorflow/tensorflow/go@r2.4-go
```

You can also use [tfgo](https://github.com/galeone/tfgo) that depends on the fork and it allows a simplified usage of the Go bindings:

```
go get github.com/galeone/tfgo
```

**Comment by gg187on** [comment]
@galeone getting error:

```
go: downloading github.com/galeone/tensorflow v2.4.0-rc0.0.20210202175351-640a390c2283+incompatible
# github.com/galeone/tensorflow/tensorflow/go
/usr/bin/ld: cannot find -ltensorflow
collect2: error: ld returned 1 exit status
```

when trying `go get github.com/galeone/tensorflow/tensorflow/go@r2.4-go`

**Comment by galeone** [comment]
> @galeone getting error:
> 
> ```
> go: downloading github.com/galeone/tensorflow v2.4.0-rc0.0.20210202175351-640a390c2283+incompatible
> # github.com/galeone/tensorflow/tensorflow/go
> /usr/bin/ld: cannot find -ltensorflow
> collect2: error: ld returned 1 exit status
> ```
> 
> when trying `go get github.com/galeone/tensorflow/tensorflow/go@r2.4-go`

This solution is old, I'm sorry.

The solution is to NOT use the official tensorflow go package, but instead use my fork:

```
go env -w GONOSUMDB="github.com/galeone/tensorflow"
```

Obviously, you should fir install the TensorFlow C library in your system (see here: https://github.com/galeone/tfgo#tensorflow-installation)

I highly recommend you to fallback on the galeone/tensorflow + galeone/tfgo solution, which is used by several go devs and it seems to work pretty well, solving all these problems.

In short:

1. Install the TensorFlow C library
2. go env -w GONOSUMDB="github.com/galeone/tensorflow"
3. go get github.com/galeone/tfgo

And in your source code, instead of using the official tensorflow package, use 

import (
        "fmt"
        tg "github.com/galeone/tfgo"
        tf "github.com/galeone/tensorflow/tensorflow/go"
)

Have a look at the readme in the tfgo repo to find a working example and a step by step guidance: https://github.com/galeone/tfgo/

Let me know if it works for you.

**Comment by gg187on** [comment]
is this correct output?

```
# github.com/galeone/tensorflow/tensorflow/go
/usr/bin/ld: /usr/local/lib/libtensorflow.so: .dynsym local symbol at index 857 (>= sh_info of 2)
/usr/bin/ld: /usr/local/lib/libtensorflow.so: .dynsym local symbol at index 1944 (>= sh_info of 2)
/usr/bin/ld: /usr/local/lib/libtensorflow.so: .dynsym local symbol at index 2314 (>= sh_info of 2)
/usr/bin/ld: /usr/local/lib/libtensorflow.so: .dynsym local symbol at index 2502 (>= sh_info of 2)
/usr/bin/ld: /usr/local/lib/libtensorflow.so: .dynsym local symbol at index 2724 (>= sh_info of 2)
/usr/bin/ld: /usr/local/lib/libtensorflow.so: .dynsym local symbol at index 2725 (>= sh_info of 2)
go get: added github.com/galeone/tfgo v0.0.0-20210204182614-84b9a5e77f79
```

**Comment by gg187on** [comment]
it seems to work, thanks so much @galeone

**Comment by galeone** [comment]
> it seems to work, thanks so much @galeone

Great! Have you fixed the linking problem you posted before?

**Comment by gg187on** [comment]
yes, today we are trying to implement something in it, and we will see if it works,... we do team project to school where we need to create CNN for object detection (there will be folder of images, and it needs to detect object on it) and then calculate time of run both in serial and parallel running. (using gorutines with it, that I'm not sure if I know how). Could You propably help us? I can pay you some money for it.

**Comment by amyxguo** [comment]
Also curious about this

**Comment by BigCatGit** [comment]
is work
